assets/js/components/products-control/index.js

```diff
@@ -2,110 +2,91 @@
-			noItems: __(
-				"Your store doesn't have any products.",
-				'woo-gutenberg-products-block'
-			),
-			search: __(
-				'Search for products to display',
-				'woo-gutenberg-products-block'
-			),
-			selected: ( n ) =>
-				sprintf(
-					_n(
-						'%d product selected',
-						'%d products selected',
-						n,
-						'woo-gutenberg-products-block'
-					),
-					n
+/**
+ * The products control exposes a custom selector for searching and selecting
+ * products.
+ *
+ * @param {function} props.onChange  Callback fired when the selected item
+ *                                   changes
+ * @param {function} props.onSearch  Callback fired when a search is triggered
+ * @param {array}    props.selected  An array of selected products.
+ * @param {array}    props.products  An array of products to select from.
+ * @param {boolean}  props.isLoading Whether or not the products are being
+ *                                   loaded.
+ *
+ * @return {function} A functional component.
+ */
+const ProductsControl = ( {
+	onChange,
+	onSearch,
+	selected,
+	products,
+	isLoading,
+} ) => {
+	const messages = {
+		clear: __( 'Clear all products', 'woo-gutenberg-products-block' ),
+		list: __( 'Products', 'woo-gutenberg-products-block' ),
+		noItems: __(
+			"Your store doesn't have any products.",
+			'woo-gutenberg-products-block'
+		),
+		search: __(
+			'Search for products to display',
+			'woo-gutenberg-products-block'
+		),
+		selected: ( n ) =>
+			sprintf(
+				_n(
+					'%d product selected',
+					'%d products selected',
+					n,
+					'woo-gutenberg-products-block'
 				),
-			updated: __(
-				'Product search results updated.',
-				'woo-gutenberg-products-block'
+				n
 			),
-		};
-
-		return (
-			<Fragment>
-				<SearchListControl
-					className="woocommerce-products"
-					list={ list }
-					isLoading={ loading }
-					selected={ selected.map( ( id ) => find( list, { id } ) ).filter( Boolean ) }
-					onSearch={ isLargeCatalog ? this.debouncedOnSearch : null }
-					onChange={ onChange }
-					messages={ messages }
-				/>
-			</Fragment>
-		);
-	}
-}
+		updated: __(
+			'Product search results updated.',
+			'woo-gutenberg-products-block'
+		),
+	};
+	return (
+		<Fragment>
+			<SearchListControl
+				className="woocommerce-products"
+				list={ products }
+				isLoading={ isLoading }
+				selected={ selected }
+				onSearch={ onSearch }
+				onChange={ onChange }
+				messages={ messages }
+			/>
+		</Fragment>
+	);
+};
 
 ProductsControl.propTypes = {
-	/**
-	 * Callback to update the selected products.
-	 */
-	onChange: PropTypes.func.isRequired,
-	/**
-	 * The list of currently selected IDs.
-	 */
-	selected: PropTypes.array.isRequired,
+	onChange: PropTypes.func,
+	onSearch: PropTypes.func,
+	selected: PropTypes.array,
+	products: PropTypes.array,
+	isLoading: PropTypes.bool,
+};
+
+ProductsControl.defaultProps = {
+	onChange: () => undefined,
+	onSearch: () => undefined,
```

Comment:
I don't know if there is any performance difference, but couldn't this be `null` instead of `() => undefined`? We already [check for `onSearch` type](https://github.com/woocommerce/woocommerce-admin/blob/master/packages/components/src/search-list-control/index.js#L53).

It shouldn't make any difference now, but at some point `<SearchListControl>`'s implementation might change and it might set the button to `disabled` if `onSearch` is falsy or something like that. In that case passing a function like this would bring unexpected results.

Reply:
In the `<SearchListControl>` component is this:

https://github.com/woocommerce/woocommerce-admin/blob/ca66af481684642cae88da79706d33b978c9bf49/packages/components/src/search-list-control/index.js#L290-L298

`onChange` is listed as requiring a function and `onSearch` is optional.  `<ProductsControl>` is essentially a wrapper around `<SearchListControl>` (which get's enhanced by `withSearchedProducts`).  So the question becomes whether we want to enforce the requirements as the wrapped component (and would probably want to keep in sync with whatever changes happen in the SearchListControl) or we don't enforce any checks on this component and allow the flow to go through to SearchListControl to throw any errors.

I think the former option would be a bit preferable so I'll implement a required check on `onChange` for ProductsControl . I believe it's preferable because it's better to fail as early as possible for consuming code to have an awareness of what needs fixed (i.e. point to incorrect usage of `<ProductsControl>` rather than `<SearchListControl>` which in this case is a hidden implementation.  I'll make that change.

As for _why_ I implemented things this way in the first place - I was thinking there might be usecases where consuming code might simply not pass in any callbacks on `onChange` or `onSearch` and thus this would allow for implementing the component without them.  But in hindsight that seems like a very edgecase (and non-useful) implementation :).

Reply:
changed in a3f49ce